A76-ring-hash-improvements.md

```diff
@@ -0,0 +1,176 @@
+endpoint hash key explicitly via EDS by reusing the configuration mechanism
+implemented in Envoy. To retain the advantage of being able to use `ring_hash`
+without xDS, custom gRPC name resolvers will be able to set this endpoint
+attribute through the language specific resolver attribute interface.
+
+### Related Proposals:
+
+This proposal extends the following existing gRFC:
+
+* [gRFC A42: xDS Ring Hash LB Policy][A42]
+
+## Proposal
+
+### Explicitly setting the request hash key
+
+A new string field `request_metadata_key` will be added to the `ring_hash`
+policy config. The ring hash policy will be modified as follows:
+
+Upon loading the load balancing config, if the `request_metadata_key` field
+contains a value that is not a valid metadata key, then the configuration is
+rejected. If the `request_metadata_key` refers to a binary metadata (prefixed
+with `-bin`), the configuration is also rejected.
+
+At pick time:
+- If `request_metadata_key` is not empty, and the request metadata has a
+non-empty value, then the request hash key will be set to this value. If
+`request_metadata_key` contains multiple values, then values are joined with a
+coma `,` character between each value before hashing.
+- If `request_metadata_key` is not empty, and the request has no value or an
+empty value associated with the metadata key defined, then the picker will
+generate a random hash for it. The use of a random hash key will effectively
+sends the request to a random endpoint.
+- If `request_metadata_key` is not set or empty, then the request hash key will
+be based on the xDS hash policy in RDS, which keeps the existing LB
+configuration for ring hash working as before with xDS.
+- If `request_metadata_key` is not set or empty but the xDS configuration does
+not provide the hash key, then the picker will generate a random hash for it to
+select a random endpoint, which matches the current behavior for xDS.
+
+
+### Explicitly setting the endpoint hash key
+
+The `ring_hash` policy will be changed such that the hash key used for
+determining the locations of each endpoint on the ring will be extracted from a
+pre-defined name resolver attribute called `hash_key`. If this attribute is set,
+then the endpoint is placed on the ring by hashing its value. If this attribute
+is not set or empty, then the endpoint IP address is used (current
+behavior). The location of an existing endpoint on the ring changes if its
+`hash_key` name resolver attribute changes.
+
+The xDS resolver will be changed so that when converting EDS responses to
+resolver endpoints, it will set the `hash_key` name resolver attribute to the
+value of [LbEndpoint.Metadata][LbEndpoint.Metadata] `envoy.lb` `hash_key` field,
+as described in [Envoy's documentation for the ring hash load
+balancer](envoy-ring-hash).
+
+### LB Policy Config Changes
+
+After the addition of this field, the `ring_hash` LB policy config will be:
+
+    message RingHashLoadBalancingConfig {
```

Comment:
To get protobuf syntax highlighting, instead of indenting the code, use a [fenced code block](https://www.markdownguide.org/extended-syntax/#syntax-highlighting), with the language name `proto`.

Reply:
done.